paddle/fluid/pir/transforms/onednn/cpu_special_ops_bf16_pass.cc

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/pir/transforms/onednn/cpu_bfloat16_squash_pass.h"
+
+#include "paddle/fluid/pir/dialect/kernel/ir/kernel_type.h"
+#include "paddle/fluid/pir/dialect/operator/ir/onednn_op.h"
+#include "paddle/fluid/pir/dialect/operator/ir/pd_op.h"
+#include "paddle/fluid/pir/drr/include/drr_pattern_base.h"
+#include "paddle/fluid/pir/utils/general_functions.h"
+
+#include "paddle/pir/include/core/builtin_op.h"
+#include "paddle/pir/include/pass/pass.h"
+#include "paddle/pir/include/pass/pass_registry.h"
+
+namespace {
+
+template <class IrType1, class IrType2>
+static pir::Type create_type(pir::Type type,
+                             pir::Type out_dtype,
+                             pir::IrContext *ctx) {
+  auto input_type = type.dyn_cast<IrType1>();
+  return IrType2::get(ctx,
+                      out_dtype,
+                      input_type.dims(),
+                      input_type.data_layout(),
+                      input_type.lod(),
+                      input_type.offset());
+}
+
+// For ops like conv and concat, their input is sometimes packed as VectorType,
+// hence current quantization doesn't work. Here we deal with them specifically.
+class ConcatBf16QuantizePattern
+    : public pir::OpRewritePattern<paddle::onednn::dialect::ConcatOp> {
+ public:
+  using pir::OpRewritePattern<
+      paddle::onednn::dialect::ConcatOp>::OpRewritePattern;
+  bool MatchAndRewrite(
+      paddle::onednn::dialect::ConcatOp op,
+      pir::PatternRewriter &rewriter) const override {  // NOLINT
+    // The input should come from combine.
+    pir::CombineOp pre_op =
+        pir::GetDefiningOpForInput(op, 0)->dyn_cast<pir::CombineOp>();
+    if (!pre_op) return false;
+    if (!pre_op.out().HasOneUse()) return false;
+
+    auto op_attributes = op->attributes();
+    auto onednn_data_type = op_attributes.at("mkldnn_data_type")
+                                .dyn_cast<pir::StrAttribute>()
+                                .AsString();
+    if (onednn_data_type == "bfloat16") return false;
```

Comment:
Also need check int8.
And btw, in cpu_bf16_placement pass, we have a lot of logic to determine whether bf16 can be used, I think we also need to add such condition for this op?

Reply:
I think it's fine for now, cause we don't have any int8 pass. And for logic about determining usability of bf16, would you mind show some examples?

Reply:
For example, only float input can be converted to bfloat16, we need to avoid int64/int32 and others use bf16 pass, and if the bfloat16 operator that is between two float32 operators, may need to revert to fp32. I'm not sure all these condition need match as cpu_bl16_placement_pass, but I think part of them should use on this op~

Reply:
Thanks~ I've already thought about 1st condition. But for 2nd one, it will be little complex, since if there is subgraph like [f32_op+f32_op+f32_concat+f32_op], the 2nd op & concat might be impacted by each other, causing both of which unable to be quantized. What about putting this to next step? Since Fused_conv might be impacted also~